### PR TITLE
[WIP] add hrnet V1 and V2

### PR DESCRIPTION
**Pull request by chinakook**
This model is all the same with the original implementation.
After this PR, I will convert the original model to gluon model.

### CODE REVIEW

**Comment by zhreshold** [comment]
Any update?

**Comment by chinakook** [comment]
@zhreshold I'll update later when I'm free.

**Comment by chinakook** [comment]
@zhreshold It's finished, please review. The pretrained model will be uploaded later after converted from py*ch.

File: gluoncv/model_zoo/hrnet.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+                    else:
+                        raise NotImplementedError
+
+                else:
+                    y = y + self.fuse_layers[i][j](X[j])
+            x_fuse.append(self.relu(y))
+
+        return x_fuse   
+
+BN_MOMENTUM = 0.9
```
Comment (zhreshold) [inline gluoncv/model_zoo/hrnet.py]: no usage of this variable in this file?

File: gluoncv/model_zoo/hrnet.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+    #modules, block_type, blocks, channels, fuse_method
+    (1,    'BOTTLENECK', [4],           [64],           'SUM'),
+    (1,    'BASIC',      [4]*2,         [48,96],        'SUM'),
+    (4,    'BASIC',      [4]*3,         [48,96,192],    'SUM'),
+    (3,    'BASIC',      [4]*4,         [48,96,192,384],'SUM')
+]
+
+def get_hrnet(model_name, stage_interp_type='nearest', purpose='cls', pretrained=False, ctx=cpu(),
+                 root='~/.mxnet/models', norm_layer=BatchNorm, norm_kwargs=None, num_classes=1000, **kwargs):
+    r"""MobileNet model from the
```
Comment (zhreshold) [inline gluoncv/model_zoo/hrnet.py]: the docstring is wrong

File: gluoncv/model_zoo/hrnet.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+def hrnet_w32_c(**kwargs):
+    return get_hrnet('w32', **kwargs)
+
+def hrnet_w40_c(**kwargs):
+    return get_hrnet('w40', **kwargs)
+
+def hrnet_w44_c(**kwargs):
+    return get_hrnet('w44', **kwargs)
+
+def hrnet_w48_c(**kwargs):
```
Comment (zhreshold) [inline gluoncv/model_zoo/hrnet.py]: so, I suppose you are trying to use `c` or `s` to separate classification or segmentation models right?
It's very confusing from the name itself, maybe we need better naming for them

File: gluoncv/model_zoo/hrnet.py (hunk 1 above)
Comment (chinakook) [reply gluoncv/model_zoo/hrnet.py]: OK

File: gluoncv/model_zoo/hrnet.py (hunk 3 above)
Comment (chinakook) [reply gluoncv/model_zoo/hrnet.py]: I looked at paddlepaddle's name scheme, I think I can add some comment. Do you have better idea?

**Comment by chinakook** [comment]
@zhreshold

**Comment by zhreshold** [comment]
@chinakook Have you tested it on COCO? With faster or mask rcnn heads?

**Comment by chinakook** [comment]
@zhreshold  I've only tested it compared with the original pytorch implementation, it got all the same results as the original implementation got. I'll release these test codes later in a new repo.
[pytorch classification implementation](https://github.com/HRNet/HRNet-Image-Classification)
[pytorch segmentation implementation](https://github.com/HRNet/HRNet-Semantic-Segmentation)

**Comment by tc-qaq** [comment]
In hrnet.py has some error for use(mxnet-cu101==1.5.0):
 1. line599,605,611 "norm_kwargs={'momentum': 0.99}, num_classes=19, **kwargs)" var of num_classes should be remove;
 2. line510,511,512 "F.contrib.BilinearResize2D(y_list[1], like=y_list[0], mode='like', align_corners=False)" var of align_corners should be remove

and this impl is unfridly to finetune seg from cls, maybe you can add basebone instead of (conv1-bn1....)

**Comment by chinakook** [comment]
> 
> 
> In hrnet.py has some error for use(mxnet-cu101==1.5.0):
> 
>     1. line599,605,611 "norm_kwargs={'momentum': 0.99}, num_classes=19, **kwargs)" var of num_classes should be remove;
> 
>     2. line510,511,512 "F.contrib.BilinearResize2D(y_list[1], like=y_list[0], mode='like', align_corners=False)" var of align_corners should be remove
> 
> 
> and this impl is unfridly to finetune seg from cls, maybe you can add basebone instead of (conv1-bn1....)

It's all the same with the original pytorch implementation. please update mxnet to 1.6.0

**Comment by chinakook** [comment]
Temporally close. After full tests, I will reopen this PR.

**Comment by chinakook** [comment]
It's the test code below, it get all the same result with the original implementation. Save the script to the [original implementation](https://github.com/HRNet/HRNet-Image-Classification) to run it.
Use this script, we can also convert the pytorch's model to mxnet. I'll do it after this PR is merged.
```python
import mxnet as mx
from mxnet import gluon
from mxnet.gluon import nn
import gluoncv

from gluoncv import model_zoo
from gluoncv.model_zoo import get_model

import numpy as np

import argparse
import torch
import torchsummary
from lib.config import config, update_config

import sys
sys.path.insert(0, 'lib/models')

from cls_hrnet import get_cls_net

parser = argparse.ArgumentParser()
parser.add_argument('--cfg',
                    help='experiment configure file name',
                    type=str, default='experiments/cls_hrnet_w18_small_v1_sgd_lr5e-2_wd1e-4_bs32_x100.yaml')
parser.add_argument('--modelDir',
                    help='model directory',
                    type=str,
                    default='')
parser.add_argument('--logDir',
                    help='log directory',
                    type=str,
                    default='')
parser.add_argument('--dataDir',
                    help='data directory',
                    type=str,
                    default='')
parser.add_argument('--testModel',
                    help='testModel',
                    type=str,
                    default='')
args = parser.parse_args()

update_config(config, args)

model = get_cls_net(config)
model.eval()

expected_image_shape = (3, 224, 224)
pt_x = torch.autograd.Variable(torch.rand(1, *expected_image_shape))
pt_y = model.forward(pt_x)
# print(pt_y)

pt_total_param = 0
S0 = []
pt_params = []
for k, v in model.named_parameters(recurse=True):
    # print(k, v.numel(), v.size())
    S0.append(v.size())
    pt_params.append(v)
    pt_total_param += v.numel()


net = model_zoo.hrnet_w18_small_v1_c()

ctx = mx.gpu(0)

net.collect_params().initialize(ctx=ctx)
x = mx.nd.random.uniform(shape=(1,3,224,224), ctx=ctx)

y = net(x)

mx_total_param = 0
mx_bias_param = 0
ig_total = 0
S1 = []

i = 0
for k,v  in net.collect_params().items():
    # print(k)
    numel = np.prod(np.array(list(v.shape)))
    if k.endswith('running_mean') or k.endswith('_running_var'):
        # print("ignore", v.shape)
        ig_total += numel
        continue
    # if k.startswith('conv') and k.endswith('bias'):
    #     mx_bias_param += numel
    # print(k, v.shape)
    S1.append(v.shape)
    d = pt_params[i].detach().numpy()
    v.set_data(d)
    mx_total_param += np.prod(numel)
    i += 1


np_x = pt_x.numpy()

mx_x = mx.nd.array(np_x, ctx=ctx)
mx_y = net(mx_x)
# print(mx_y)

mx_pt_y = mx.nd.array(pt_y.detach().numpy(), ctx=ctx)
mx_pt_y = mx_pt_y.as_in_context(mx.cpu())

mx_y = mx_y.as_in_context(mx.cpu())
assert np.allclose(mx_y.asnumpy(), pt_y.detach().numpy(), 1e-3)

```

**Comment by zhreshold** [comment]
@chinakook Thanks, I can merge once the lint error is fixed.

**Comment by mli** [comment]
Job PR-1081-11 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1081/11/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@chinakook Thanks, this is merged, let me know if you have follow ups ready!
